task_sdk/src/airflow/sdk/execution_time/supervisor.py

```diff
@@ -391,12 +391,55 @@ def _send_startup_message(self, ti: TaskInstance, path: str | os.PathLike[str],
         self.stdin.write(msg.model_dump_json().encode())
         self.stdin.write(b"\n")
 
-    def kill(self, signal: signal.Signals = signal.SIGINT):
+    def kill(
+        self,
+        signal_to_send: signal.Signals = signal.SIGINT,
+        escalation_delay: float = 5.0,
+        bypass_escalation: bool = False,
+    ):
+        """
+        Attempt to terminate the subprocess with a given signal.
+
+        If the process does not exit within `escalation_delay` seconds, escalate to SIGTERM and eventually SIGKILL if necessary.
+
+        :param signal_to_send: The signal to send initially (default is SIGINT).
+        :param escalation_delay: Time in seconds to wait before escalating to a stronger signal.
+        :param bypass_escalation: If True, send the signal directly to the process without escalation.
+        """
         if self._exit_code is not None:
             return
 
-        with suppress(ProcessLookupError):
-            os.kill(self.pid, signal)
+        if bypass_escalation:
+            with suppress(ProcessLookupError):
+                os.kill(self.pid, signal_to_send)
+            return
+
+        # Escalation sequence: SIGINT -> SIGTERM -> SIGKILL
+        escalation_path = [signal.SIGINT, signal.SIGTERM, signal.SIGKILL]
+        if signal_to_send in escalation_path:
+            # Start from `initial_signal`
+            escalation_path = escalation_path[escalation_path.index(signal_to_send) :]
+
+        for sig in escalation_path:
+            try:
+                if sig == signal.SIGKILL:
+                    self._process.kill()
+                elif sig == signal.SIGTERM:
+                    self._process.terminate()
+                else:
+                    os.kill(self.pid, sig)
```

Comment:
Do we need to handle this differently, or can we just use `os.kill` always?

Reply:
We could get some free goodies by directly calling `process.{kill,terminate}` -- some additional checks and cross-platform :

https://github.com/giampaolo/psutil/blob/master/psutil/__init__.py#L1326-L1334

```py
    def kill(self):
        """Kill the current process with SIGKILL pre-emptively checking
        whether PID has been reused.
        """
        if POSIX:
            self._send_signal(signal.SIGKILL)
        else:  # pragma: no cover
            self._raise_if_pid_reused()
            self._proc.kill()
```

Reply:
In which case, use `self._process.send_singal(sig)` -- that does the same https://github.com/giampaolo/psutil/blob/master/psutil/__init__.py#L1278-L1291

Reply:
oh nice, done in https://github.com/apache/airflow/pull/44465/commits/e6105241a0d56d4fac90cdfe96e75e379551ebd1